File: app/static/js/elgamal.js
// ElGamal 加密算法 - 浏览器版本
class ElGamal {
    constructor(bits) {
        this.bits = bits;
        this.isCleaned = false;
        this.p = null;
        this.g = null;
        this.y = null;
        this.x = null;
        this.q = null;
        
        // 初始化随机数生成器
        this.initRandom();
    }
    
    destructor() {
        if (!this.isCleaned) {
            this.clean();
        }
    }
    
    // 初始化随机数生成器（浏览器版本）
    initRandom() {
        // 检查是否支持 crypto.getRandomValues
        if (typeof window !== 'undefined' && window.crypto && window.crypto.getRandomValues) {
            this.cryptoRandom = window.crypto;
        } else {
            console.warn('Crypto API not available, using Math.random() (not cryptographically secure)');
            this.cryptoRandom = null;
        }
    }
    
    // 生成随机字节（浏览器版本）
    randomBytes(length) {
        const array = new Uint8Array(length);
        if (this.cryptoRandom) {
            this.cryptoRandom.getRandomValues(array);
        } else {
            // 回退到 Math.random()
            for (let i = 0; i < length; i++) {
                array[i] = Math.floor(Math.random() * 256);
            }
        }
        return array;
    }
    
    // 检查是否为素数 (Miller-Rabin 素性测试)
    isPrime(n, k = 10) {
        if (n === 2n || n === 3n) return true;
        if (n < 2n || n % 2n === 0n) return false;
        
        // 写 n-1 为 d * 2^r
        let d = n - 1n;
        let r = 0;
        while (d % 2n === 0n) {
            d /= 2n;
            r++;
        }
        
        // Miller-Rabin 测试
        for (let i = 0; i < k; i++) {
            const a = this.randomBigInt(2n, n - 2n);
            let x = this.modPow(a, d, n);
            
            if (x === 1n || x === n - 1n) continue;
            
            let composite = true;
            for (let j = 0; j < r - 1; j++) {
                x = this.modPow(x, 2n, n);
                if (x === n - 1n) {
                    composite = false;
                    break;
                }
            }
            
            if (composite) return false;
        }
        
        return true;
    }
    
    // 生成指定位数的随机素数
    getPrime(bits) {
        while (true) {
            const candidate = this.randomBigInt(2n ** BigInt(bits - 1), 2n ** BigInt(bits) - 1n);
            // 确保是奇数
            const oddCandidate = candidate | 1n;
            if (this.isPrime(oddCandidate)) {
                return oddCandidate;
            }
        }
    }
    
    // 生成安全素数 p = 2q + 1
    genSafePrime(bits) {
        while (true) {
            // 生成素数 q
            const q = this.getPrime(bits - 1);
            const p = 2n * q + 1n;
            
            // 检查 p 是否为素数
            if (this.isPrime(p)) {
                return { p, q };
            }
        }
    }
    
    // 生成指定范围内的随机大整数（修改为使用浏览器API）
    randomBigInt(min, max) {
        const range = max - min + 1n;
        const bitLength = range.toString(2).length;
        const byteLength = Math.ceil(bitLength / 8);
        
        while (true) {
            const randomBytes = this.randomBytes(byteLength);
            let randomBigInt = 0n;
            
            for (let i = 0; i < randomBytes.length; i++) {
                randomBigInt = (randomBigInt << 8n) + BigInt(randomBytes[i]);
            }
            
            randomBigInt = randomBigInt % range;
            const result = min + randomBigInt;
            
            if (result <= max) {
                return result;
            }
        }
    }
    
    // 快速模幂运算
    modPow(base, exponent, modulus) {
        if (modulus === 1n) return 0n;
        
        let result = 1n;
        base = base % modulus;
        
        while (exponent > 0n) {
            if (exponent % 2n === 1n) {
                result = (result * base) % modulus;
            }
            exponent = exponent >> 1n;
            base = (base * base) % modulus;
        }
        
        return result;
    }
    
    // 扩展欧几里得算法求模逆
    modInverse(a, m) {
        if (m === 1n) return 0n;
        
        const m0 = m;
        let x0 = 0n, x1 = 1n;
        
        while (a > 1n) {
            const q = a / m;
            let t = m;
            
            m = a % m;
            a = t;
            t = x0;
            
            x0 = x1 - q * x0;
            x1 = t;
        }
        
        if (x1 < 0n) x1 += m0;
        return x1;
    }
    
    // 生成密钥对
    keygen() {
        // 1. 生成素数 p = 2q + 1
        const { p, q } = this.genSafePrime(this.bits);
        this.p = p;
        this.q = q;
        
        // 2. 选取生成元 g
        while (true) {
            // 生成随机数 h ∈ [2, p-1]
            const h = this.randomBigInt(2n, this.p - 1n);
            
            // g = h^2 mod p
            this.g = this.modPow(h, 2n, this.p);
            if (this.g > 1n) {
                break;
            }
        }
        
        // 3. 生成私钥 x ∈ [1, q-1]
        this.x = this.randomBigInt(1n, this.q - 1n);
        
        // 4. 计算公钥 y = g^x mod p
        this.y = this.modPow(this.g, this.x, this.p);
    }
    
    // 生成私钥
    generatePrivateKey() {
        this.q = (this.p - 1n) / 2n;
        
        // 生成私钥 x ∈ [1, q-1]
        this.x = this.randomBigInt(1n, this.q - 1n);
    }
    
    // 设置公钥参数
    setPKG(p, g, y) {
        this.p = BigInt(p);
        this.g = BigInt(g);
        this.y = BigInt(y);
    }
    
    // 初始化 x
    initX() {
        this.x = this.randomBigInt(1n, this.p - 1n);
    }
    
    // 获取公钥参数
    getPKG() {
        return {
            p: this.p,
            g: this.g,
            y: this.y
        };
    }
    
    // 加密消息
    encrypt(m) {
        m = BigInt(m);
        
        try {
            this.checkM(m);
        } catch (error) {
            throw error;
        }
        
        // 生成随机数 k ∈ [1, q-1]
        const k = this.randomBigInt(1n, this.q - 1n);
        
        // c1 = g^k mod p
        const c1 = this.modPow(this.g, k, this.p);
        
        // c2 = m * y^k mod p
        const c2 = (m * this.modPow(this.y, k, this.p)) % this.p;
        
        return { c1, c2 };
    }
    
    // 解密消息
    decrypt(c1, c2) {
        c1 = BigInt(c1);
        c2 = BigInt(c2);
        
        // s = c1^x mod p
        const s = this.modPow(c1, this.x, this.p);
        
        // m = c2 * s^(-1) mod p
        const sInv = this.modInverse(s, this.p);
        const m = (c2 * sInv) % this.p;
        
        return m;
    }
    
    // 清理资源
    clean() {
        this.p = null;
        this.g = null;
        this.y = null;
        this.x = null;
        this.q = null;
        this.isCleaned = true;
    }
    
    // 生成随机消息 m ∈ [2, q-1]
    getM() {
        return this.randomBigInt(2n, this.q - 1n);
    }
    
    // 检查消息是否有效
    checkM(m) {
        if (m >= this.p || m < 2n) {
            throw new Error("Invalid message: m must be in [2, p-1]");
        }
    }
}

// 浏览器环境下的全局暴露
if (typeof window !== 'undefined') {
    window.ElGamal = ElGamal;
}

File: utils/elgamal/elgamal.py
import random
import time
from Crypto.Util import number
from Crypto.Random import random as crypto_random

class ElGamal:
    def __init__(self, bits = 256):
        self.bits = bits
        self.is_cleaned = False
        self.p = None
        self.g = None
        self.y = None
        self.x = None
        self.q = None
        
        # 初始化随机数生成器
        seed = int(time.time() * 1000000) ^ id(self) ^ random.getrandbits(32)
        random.seed(seed)
    
    def __del__(self):
        if not self.is_cleaned:
            self.clean()
    
    def gen_safe_prime(self, bits):
        """生成安全素数 p = 2q + 1"""
        while True:
            # 生成素数 q
            q = number.getPrime(bits - 1)
            p = 2 * q + 1
            
            # 检查 p 是否为素数
            if number.isPrime(p):
                return p, q
    
    def keygen(self):
        """生成密钥对"""
        # 1. 生成素数 p = 2q + 1
        self.p, self.q = self.gen_safe_prime(self.bits)
        
        # 2. 选取生成元 g
        while True:
            # 生成随机数 h ∈ [2, p-1]
            h = crypto_random.randrange(2, self.p)
            
            # g = h^2 mod p
            self.g = pow(h, 2, self.p)
            if self.g > 1:
                break
        
        # 3. 生成私钥 x ∈ [1, q-1]
        self.x = crypto_random.randrange(1, self.q)
        
        # 4. 计算公钥 y = g^x mod p
        self.y = pow(self.g, self.x, self.p)
    
    def generate_private_key(self):
        """生成私钥"""
        self.q = (self.p - 1) // 2
        
        # 生成私钥 x ∈ [1, q-1]
        self.x = crypto_random.randrange(1, self.q)
        
        # 刷新公钥 y
        self.y = pow(self.g, self.x, self.p)
    
    def set_pkg(self, p, g, y):
        """设置公钥参数"""
        self.p = p
        self.g = g
        self.y = y
    
    def init_x(self):
        """初始化 x"""
        self.x = crypto_random.randrange(1, self.p)
    
    def get_pkg(self):
        """获取公钥参数"""
        return self.p, self.g, self.y
    
    def encrypt(self, m):
        """加密消息"""
        try:
            self.check_m(m)
        except ValueError as e:
            raise e
        
        # 生成随机数 k ∈ [1, q-1]
        k = crypto_random.randrange(1, self.q)
        
        # c1 = g^k mod p
        c1 = pow(self.g, k, self.p)
        
        # c2 = m * y^k mod p
        c2 = (m * pow(self.y, k, self.p)) % self.p
        
        return c1, c2
    
    def decrypt(self, c1, c2):
        """解密消息"""
        # s = c1^x mod p
        s = pow(c1, self.x, self.p)
        
        # m = c2 * s^(-1) mod p
        s_inv = pow(s, -1, self.p)
        m = (c2 * s_inv) % self.p
        
        return m
    
    def clean(self):
        """清理资源"""
        self.p = None
        self.g = None
        self.y = None
        self.x = None
        self.q = None
        self.is_cleaned = True
    
    def get_m(self):
        """生成随机消息 m ∈ [2, q-1]"""
        return crypto_random.randrange(2, self.q)
    
    def check_m(self, m):
        """检查消息是否有效"""
        if m >= self.p or m < 1:  
            raise ValueError(f"Invalid message: m must be in [1, p-1], got {m}")
        if m == 0:
            raise ValueError("Message cannot be 0")

# 使用示例
if __name__ == "__main__":
    # 创建 ElGamal 实例
    elgamal = ElGamal(512)  # 512 位
    
    # 生成密钥
    elgamal.keygen()
    
    # 打印公钥参数
    p, g, y = elgamal.get_pkg()
    print(f"p = {p}")
    print(f"g = {g}")
    print(f"y = {y}")
    
    # 生成消息
    message = elgamal.get_m()
    print(f"Original message: {message}")
    
    # 加密
    c1, c2 = elgamal.encrypt(message)
    print(f"Encrypted: c1 = {c1}, c2 = {c2}")
    
    # 解密
    decrypted = elgamal.decrypt(c1, c2)
    print(f"Decrypted message: {decrypted}")
    
    # 验证
    print(f"Encryption/Decryption successful: {message == decrypted}")